include/click/batchelement.hh

```diff
@@ -131,5 +131,127 @@ class BatchElement : public Element { public:
 };
 #endif
 
+/**
+ * Batch helper element.
+ *
+ * Implement a (batch-compatible) element only by providing a function that
+ *  returns a port index. It uses CRTP so no virtual call is added.
+ *
+ * Inherited class must implement int classify(Packet* p);
+ *
+ * The inherited element cannot be extended further because of CRTP !
+ */
+
+template <typename T>
+class ClassifyElement : public BatchElement { public:
+    void push(int, Packet *p)
+    {
+        checked_output_push(static_cast<T&>(*this).classify(p),p);
+    }
+
+#if HAVE_BATCH
+    void push_batch(int, PacketBatch *batch)
+    {
+          CLASSIFY_EACH_PACKET(noutputs() + 1, static_cast<T&>(*this).classify,batch,checked_output_push_batch);
+    }
+#endif
+
+    private:
+        ClassifyElement(){};
+        friend T;
+};
+
+
+/**
+ * Batch helper element.
+ *
+ * Allows the user to simply implement simple_action and simple_action_batch
+ *  but relying on CRTP instead of virtual, removing one virtual call.
+ *
+ * The inherited element cannot be extended further because of CRTP !
```

Comment:
What countermeasures do we have for this restriction?
If we convert all Click elements to this template, how can we leverage inheritance?
Maybe using template parameters, but this makes the code base more complicated.

Reply:
This is a fast easy way to make an element compatible. If one wants inheritance he can copy paste the code in the CRTP. Or extend the base template to pass the inherited and the final inheritance. But in any case if someone wants to do something complex he'll implement push* and pull* himself...